{
 "metadata": {
  "name": "",
  "signature": "sha256:9554a488878cb5aa48214f8562e25be729adcb59b1240185f7c99574cadf6f9a"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "from __future__ import division\n",
      "%matplotlib inline\n",
      "import os\n",
      "import math\n",
      "import numpy as np\n",
      "import matplotlib.pyplot as plt\n",
      "import pylab\n",
      "from matplotlib import rc\n",
      "import scipy as sc\n",
      "import pandas as pd\n",
      "import scipy.stats as stats\n",
      "from clean_data import CANCER_TYPES\n",
      "\n",
      "from IPython.html.widgets import interact\n",
      "from IPython.html import widgets\n",
      "from IPython.display import display\n",
      "\n",
      "from mutation_analysis import get_pancancer_mutation_summary\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Correlation Between Silent and Non-Silent Mutations\n",
      "- Count number of silent and non-silent mutations for every SAMPLE across all genes and compute correlation\n",
      "- Count number of silent and non-silent mutations for every GENE across samples and compute correlation"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "can_types = []\n",
      "for can in CANCER_TYPES:\n",
      "    f1 = '../data/processed/' + can + os.sep + can + '_nsilent_mutation.txt'\n",
      "    f2 = '../data/processed/' + can + os.sep + can + '_silent_mutation.txt'\n",
      "    if os.path.exists(f1) and os.path.exists(f2):\n",
      "        can_types.append(can)\n",
      "\n",
      "print \"There are %d cancer types ready to be analysed\" % len(can_types)\n",
      "\n",
      "@interact(cancer_type=widgets.DropdownWidget(values=can_types), Gene_or_Sample=widgets.DropdownWidget(values={'GENE':1, 'SAMPLE':0}))\n",
      "def plot_corr(cancer_type, Gene_or_Sample):\n",
      "    can = cancer_type\n",
      "    f1 = '../data/processed/' + can + os.sep + can + '_nsilent_mutation.txt'\n",
      "    nsdf = pd.read_table(f1, header=0, index_col=0)\n",
      "    nsilent = nsdf.sum(axis=Gene_or_Sample)\n",
      "\n",
      "    f2 = '../data/processed/' + can + os.sep + can + '_silent_mutation.txt'\n",
      "    sdf = pd.read_table(f2, header=0, index_col=0)\n",
      "    silent = sdf.sum(axis=Gene_or_Sample)\n",
      "\n",
      "\n",
      "    \n",
      "    print \"# Total non-silent mutation: %d, # total silent mutation: %d\" %(sum(nsilent), sum(silent))\n",
      "    print \"# Per SAMPLE non-silent mutation: %f, # silent mutation: %f\" %(sum(nsilent)/sdf.shape[1], sum(silent)/sdf.shape[1])\n",
      "    print \"# Per GENE non-silent mutation: %f, # silent mutation: %f\" %(sum(nsilent)/nsdf.shape[0], sum(silent)/nsdf.shape[0])\n",
      "    print nsdf.shape, sdf.shape\n",
      "\n",
      "    total = silent + nsilent\n",
      "    \n",
      "    if Gene_or_Sample:\n",
      "        txt = 'GENE'\n",
      "    else:\n",
      "        txt = 'SAMPLE'\n",
      "            \n",
      "    pylab.rcParams['figure.figsize'] = (12.0, 8.0)\n",
      "    plt.subplot(3, 1, 1)\n",
      "    plt.plot(range(1, len(silent)+1), silent/sdf.count(axis=Gene_or_Sample), label='Silent Mutations')\n",
      "    plt.plot(range(1, len(nsilent)+1), nsilent/nsdf.count(axis=Gene_or_Sample), label='Non-Silent Mutation')\n",
      "    plt.ylabel(\"Mutation Frequency\")\n",
      "    plt.title(can + \" silent vs non-silent mutations for every \" + txt + \" Spearman's corr.: \" + str(silent.corr(nsilent, 'spearman')) + \"\\n\")\n",
      "    plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
      "#     plt.yscale('log')\n",
      "    plt.grid()\n",
      "    \n",
      "    # High frequency mutations\n",
      "    plt.subplot(3, 1, 2)\n",
      "    nhigh_freq = nsilent.order(ascending=False).iloc[:20]*100/nsdf.shape[int(Gene_or_Sample)]\n",
      "    nhigh_freq.plot(kind='bar', rot=30)\n",
      "    plt.ylabel(\"Non-Silent Mutation Freq(%)\")\n",
      "    plt.xlabel(txt)\n",
      "    plt.grid()\n",
      "\n",
      "    # High frequency mutations silent + non-silent\n",
      "    plt.subplot(3, 1, 3)\n",
      "    thigh_freq = total.order(ascending=False).iloc[:20]*100/sdf.shape[int(Gene_or_Sample)]\n",
      "    thigh_freq.plot(kind='bar', rot=30)\n",
      "    plt.ylabel(\"Overall Mutation Freq(%)\")\n",
      "    plt.xlabel(txt)\n",
      "    plt.grid()\n",
      "\n",
      "    plt.tight_layout(pad=0.2, w_pad=0.5, h_pad=0.2)\n",
      "    #print df.shape[int(Gene_or_Sample)]\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Observations \n",
      "1. A very high correlation between the total number of silent and non-silent mutation is observed for every patient. This implies that if one considers each patient individually then it will be hard to separate drivers from passenger mutations. This appears reasonable as very often cancer messes up the DNA repair mechanisms.\n",
      "2. In contrast to the above, the gene-wise correlation between the total number of silent and non-silent mutations across patient population is weak. This suggests that different patients harobor different set of mutations. In other words, only very few genes show high frequency of mutation across patients which makes it tough to identify them.\n",
      "3. Genes with very high silent mutation freq should be neglected????\n",
      "4. Several genes have very high mutation frequency in LUAD and LUSC. Whats wrong?"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## TODO\n",
      "1. Does silent mutations have higher probability because of codon usage reduandancy? Read  Identifying cancer driver genes in tumor genome sequencing studies, Richard simon, bioinformatics\n",
      "2. Box plot mutation per gene and mutation per sample across cancer types"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "gene = 0; refresh = True    \n",
      "ns_summary, s_summary = get_pancancer_mutation_summary(gene, refresh)\n",
      "outdir = '../data/pancancer'\n",
      "if gene:\n",
      "    ns_outf = outdir + os.sep + 'non_silent_mutations_genes_all.txt'               \n",
      "    s_outf = outdir + os.sep + 'silent_mutations_genes_all.txt' \n",
      "    outf='pancancer_genewise_mutation_boxplot.png'\n",
      "else:\n",
      "    ns_outf = outdir + os.sep + 'non_silent_mutations_samples_all.txt'               \n",
      "    s_outf = outdir + os.sep + 'silent_mutations_samples_all.txt' \n",
      "    outf='pancancer_samplewise_mutation_boxplot.png'\n",
      "ns_summary = pd.read_table(ns_outf, index_col=0, header=0)\n",
      "s_summary = pd.read_table(s_outf, index_col=0, header=0)\n",
      "fig = plotly_boxplots(ns_summary, s_summary, outf)\n",
      "#     plot_url = py.plot(fig, filename='box-grouped')\n",
      "#     py.iplot(fig, filename = 'basic-line')\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "import plotly\n",
      "import plotly.plotly as py\n",
      "from plotly.graph_objs import *\n",
      "\n",
      "plotly.tools.set_credentials_file(username='extra.nitin', api_key='auykqr8hfa', stream_ids=['kvd6yka2q7', 'u6wv80rlbm'])\n",
      "\n",
      "def plotly_boxplots(ns_summary, s_summary, outf):\n",
      "    \"\"\" Plot Non-Silent and Silent Mutations box-plot across all\n",
      "    cancer types\n",
      "    \"\"\"\n",
      "    s_stacked = s_summary.stack()\n",
      "    s_stacked.index = s_stacked.index.droplevel(0)\n",
      "    s_stacked = pd.DataFrame(s_stacked).sort_index()\n",
      "    silent_x = list(s_stacked.index)\n",
      "    silent_y = list(s_stacked.values)\n",
      "\n",
      "    ns_stacked = ns_summary.stack()\n",
      "    ns_stacked.index = ns_stacked.index.droplevel(0)\n",
      "    ns_stacked = pd.DataFrame(ns_stacked).sort_index()\n",
      "    nsilent_x = list(ns_stacked.index)\n",
      "    nsilent_y = list(ns_stacked.values)\n",
      "\n",
      "    # Now plot the box plots\n",
      "    trace1 = Box(\n",
      "        y=silent_y,\n",
      "        x=silent_x,\n",
      "        name='Silent',\n",
      "        marker=Marker(\n",
      "            color='#3D9970'\n",
      "        )\n",
      "    )\n",
      "    trace2 = Box(\n",
      "        y=nsilent_y,\n",
      "        x=nsilent_x,\n",
      "        name='Non-Silent',\n",
      "        marker=Marker(\n",
      "            color='#FF4136'\n",
      "        )\n",
      "    )\n",
      "\n",
      "    data = Data([trace1, trace2])\n",
      "    layout = Layout(\n",
      "        title='Mutation Landscape of the TCGA cancer cohorts',\n",
      "        titlefont=Font(\n",
      "                size=26,\n",
      "            ),\n",
      "        xaxis=XAxis(\n",
      "            tickangle=-45,\n",
      "            tickfont=Font(\n",
      "                size=18,\n",
      "            ),\n",
      "        ),\n",
      "        yaxis=YAxis(\n",
      "            title='Number of Mutations',\n",
      "            titlefont=Font(\n",
      "                size=22,\n",
      "            ),\n",
      "            zeroline=False,\n",
      "            type='log',\n",
      "            autorange=True\n",
      "        ),\n",
      "        autosize=False,\n",
      "        width=1800,\n",
      "        height=1200,\n",
      "        boxmode='group'\n",
      "    )\n",
      "    fig = Figure(data=data, layout=layout)\n",
      "\n",
      "    pylab.rcParams['figure.figsize'] = (18.0, 12.0)\n",
      "    py.image.ishow(fig)\n",
      "    out_png = '../figures/' + outf\n",
      "    py.image.save_as(fig, out_png)\n",
      "\n",
      "    return fig\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene = 0; refresh = True    \n",
      "ns_summary, s_summary = get_pancancer_mutation_summary(gene, refresh)\n",
      "outdir = '../data/pancancer'                                                   \n",
      "ns_outf = outdir + os.sep + 'non_silent_mutations_samples_all.txt'               \n",
      "s_outf = outdir + os.sep + 'silent_mutations_samples_all.txt' \n",
      "ns_summary = pd.read_table(ns_outf, index_col=0, header=0)\n",
      "s_summary = pd.read_table(s_outf, index_col=0, header=0)\n",
      "\n",
      "\n",
      "pylab.rcParams['figure.figsize'] = (18.0, 12.0)\n",
      "fig, axes = plt.subplots(nrows=2, ncols=1)\n",
      "ns_summary.plot(ax=axes[0], kind='box', logy=True, rot=30); axes[0].set_ylabel('Number of Non-Silent Mutation in Samples')\n",
      "s_summary.plot(ax=axes[1], kind='box', logy=True, rot=30); axes[1].set_ylabel('Number of Silent Mutation in Samples');\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "import brewer2mpl\n",
      "\n",
      "outdir = '../data/pancancer'                                                   \n",
      "ns_outf = outdir + os.sep + 'non_silent_mutations_samples_all.txt'               \n",
      "s_outf = outdir + os.sep + 'silent_mutations_samples_all.txt' \n",
      "ns_summary = pd.read_table(ns_outf, index_col=0, header=0)\n",
      "s_summary = pd.read_table(s_outf, index_col=0, header=0)\n",
      "\n",
      "both_df = ns_summary.join(s_summary, lsuffix='_NS')\n",
      "\n",
      "plt.close('all')\n",
      "fig, axes = plt.subplots(nrows=5, ncols=5, figsize=(18, 18), dpi=1200)\n",
      "pylab.rcParams['figure.max_open_warning'] = 0\n",
      "set2 = brewer2mpl.get_map('Set2', 'qualitative', 8).mpl_colors\n",
      "for idx, c in enumerate(ns_summary.columns):\n",
      "    tmp_df = both_df[[c+'_NS', c]].dropna()#.fillna(0)\n",
      "    x = tmp_df[c].values\n",
      "    y = tmp_df[c+'_NS'].values\n",
      "    i = idx//5; j = idx%5\n",
      "    axes[i, j].scatter(x, y, color=set2[idx%8])\n",
      "    axes[i, j].set_xscale('log')\n",
      "    axes[i, j].set_yscale('log')\n",
      "    axes[i, j].set_title('%s %.2f, %.2f' %(c, stats.pearsonr(x, y)[0], stats.spearmanr(x,y)[0]))\n",
      "    \n",
      "    if i==2 and j == 0:\n",
      "        axes[i, j].set_ylabel('Non-Silent Mutations', size=20)\n",
      "    if i==4 and j == 2:\n",
      "        axes[i, j].set_xlabel('Silent Mutations', size=20)\n",
      "\n",
      "    axes[i, j].grid()\n",
      "# fig.suptitle('Non-Silent v/s Silent Mutations: Pearson and Spearman Correlations\\n', size=12)\n",
      "plt.tight_layout()\n",
      "plt.show()\n",
      "fig.savefig('../figures/pancancer_mutation_individual_correlation.eps', format='eps')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "import brewer2mpl\n",
      "\n",
      "outdir = '../data/pancancer'                                                   \n",
      "ns_outf = outdir + os.sep + 'non_silent_mutations_samples_all.txt'               \n",
      "s_outf = outdir + os.sep + 'silent_mutations_samples_all.txt' \n",
      "ns_summary = pd.read_table(ns_outf, index_col=0, header=0)\n",
      "s_summary = pd.read_table(s_outf, index_col=0, header=0)\n",
      "\n",
      "\n",
      "both_df = ns_summary.join(s_summary, lsuffix='_NS')\n",
      "\n",
      "# pylab.rcParams['figure.max_open_warning'] = 0\n",
      "# pylab.rcParams['figure.figsize'] = (22.0, 12.0)\n",
      "\n",
      "set2 = brewer2mpl.get_map('Set2', 'qualitative', 8).mpl_colors\n",
      "markers = ['x', 'o', '*']\n",
      "fig, ax = plt.subplots(1, figsize=(16, 14), dpi=1200)\n",
      "for i, c in enumerate(ns_summary.columns):\n",
      "    tmp_df = both_df[[c+'_NS', c]].dropna()#.fillna(0)\n",
      "    x = tmp_df[c].values\n",
      "    y = tmp_df[c+'_NS'].values\n",
      "    ax.scatter(x, y, color=set2[i%8], marker=markers[i%3], label='%s, P: %.2f, S: %.2f' %(c, stats.pearsonr(x, y)[0], stats.spearmanr(x,y)[0]))\n",
      "ax.set_xscale('log')\n",
      "ax.set_yscale('log')\n",
      "\n",
      "ax.set_xlabel('Silent Mutations', size=18)\n",
      "ax.set_ylabel('Non-Silent Mutations', size=18)\n",
      "\n",
      "# The overall correlataion\n",
      "xa = ns_summary.fillna(0).sum(1)\n",
      "ya = s_summary.fillna(0).sum(1)\n",
      "common = xa.index.intersection(ya.index)\n",
      "xa = xa[common].values\n",
      "ya = ya[common].values\n",
      "plt.title('Pancancer Non-Silent v/s Silent Mutations Correlation Pearson: %.2f, Spearman: %.2f'  %(stats.pearsonr(xa, ya)[0], stats.spearmanr(xa,ya)[0]), size=20)\n",
      "# ax.legend(loc='upper center', bbox_to_anchor=(0.5, 1.05),\n",
      "#           ncol=5, fancybox=True, shadow=True)\n",
      "# plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
      "plt.legend(loc='upper left')\n",
      "plt.grid()\n",
      "\n",
      "fig.savefig('../figures/pancancer_mutation_aggregated_correlation.eps', format='eps')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Check gene level enrichments\n",
      "Apply Fisher's exact test. **Null Hypothesis: Silent and Non silent mutations have same distribution**\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "for can in CANCER_TYPES:\n",
      "    f1 = '../data/processed/' + can + os.sep + can + '_nsilent_mutation.txt'\n",
      "    f2 = '../data/processed/' + can + os.sep + can + '_silent_mutation.txt'\n",
      "    if os.path.exists(f1) and os.path.exists(f2):\n",
      "        can_types.append(can)\n",
      "\n",
      "print \"There are %d cancer types ready to be analysed\" % len(can_types)\n",
      "\n",
      "#@interact(cancer_type=widgets.DropdownWidget(values=can_types), Gene_or_Sample=widgets.DropdownWidget(values={'GENE':1, 'SAMPLE':0}))\n",
      "def plot_corr(cancer_type, Gene_or_Sample):\n",
      "    can = cancer_type\n",
      "    f1 = '../data/processed/' + can + os.sep + can + '_nsilent_mutation.txt'\n",
      "    nsdf = pd.read_table(f1, header=0, index_col=0)\n",
      "    nsilent = nsdf.sum(axis=Gene_or_Sample)*-1\n",
      "    \n",
      "    f2 = '../data/processed/' + can + os.sep + can + '_silent_mutation.txt'\n",
      "    sdf = pd.read_table(f2, header=0, index_col=0)\n",
      "    silent = sdf.sum(axis=Gene_or_Sample)\n",
      "\n",
      "    nsamples = nsdf.shape[1]\n",
      "    return (nsilent, silent, nsamples)\n",
      "\n",
      "    \n",
      "#     print \"# Total non-silent mutation: %d, # total silent mutation: %d\" %(sum(nsilent), sum(silent))\n",
      "#     print \"# Per SAMPLE non-silent mutation: %f, # silent mutation: %f\" %(sum(nsilent)/sdf.shape[1], sum(silent)/sdf.shape[1])\n",
      "#     print \"# Per GENE non-silent mutation: %f, # total silent mutation: %f\" %(sum(nsilent)/nsdf.shape[0], sum(silent)/nsdf.shape[0])\n",
      "#     print nsdf.shape, sdf.shape\n",
      "\n",
      "#     total = silent + nsilent\n",
      "    \n",
      "#     if Gene_or_Sample:\n",
      "#         txt = 'GENE'\n",
      "#     else:\n",
      "#         txt = 'SAMPLE'\n",
      "            \n",
      "#     pylab.rcParams['figure.figsize'] = (12.0, 8.0)\n",
      "#     plt.subplot(3, 1, 1)\n",
      "#     plt.plot(range(1, len(silent)+1), silent, label='Silent Mutations')\n",
      "#     plt.plot(range(1, len(nsilent)+1), nsilent, label='Non-Silent Mutation')\n",
      "#     plt.ylabel(\"Mutated gene count\")\n",
      "#     plt.title(can + \" silent vs non-silent mutations for every \" + txt + \" Spearman's corr.: \" + str(silent.corr(nsilent, 'spearman')) + \"\\n\")\n",
      "#     plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
      "#     plt.grid()\n",
      "    \n",
      "#     # High frequency mutations\n",
      "#     plt.subplot(3, 1, 2)\n",
      "#     nhigh_freq = nsilent.order(ascending=False).iloc[:20]*100/nsdf.shape[int(Gene_or_Sample)]\n",
      "#     nhigh_freq.plot(kind='bar', rot=30)\n",
      "#     plt.ylabel(\"Non-Silent Mutation Freq(%)\")\n",
      "#     plt.xlabel(txt)\n",
      "#     plt.grid()\n",
      "\n",
      "#     # High frequency mutations silent + non-silent\n",
      "#     plt.subplot(3, 1, 3)\n",
      "#     thigh_freq = total.order(ascending=False).iloc[:20]*100/sdf.shape[int(Gene_or_Sample)]\n",
      "#     thigh_freq.plot(kind='bar', rot=30)\n",
      "#     plt.ylabel(\"Overall Mutation Freq(%)\")\n",
      "#     plt.xlabel(txt)\n",
      "#     plt.grid()\n",
      "\n",
      "#     plt.tight_layout(pad=0.2, w_pad=0.5, h_pad=0.2)\n",
      "#     print df.shape[int(Gene_or_Sample)]\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}